File: algorithms/2_medium/B/00103_binary_tree_zigzag_level_order_traversal/bfs.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
use std::collections::VecDeque;
impl Solution {
    pub fn zigzag_level_order(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
        if root.is_none() {
            return Vec::new();
        }

        let mut q: VecDeque<Rc<RefCell<TreeNode>>> = VecDeque::new();
        q.push_back(root.as_ref().unwrap().clone());
        let mut flip_flop = false;
        let mut res: Vec<Vec<i32>> = Vec::new();

        while !q.is_empty() {
            let len = q.len();
            let mut temp: Vec<i32> = Vec::with_capacity(len);

            for _ in 0..len{
                let node_rc = q.pop_front().unwrap();
                let node = node_rc.borrow();
                temp.push(node.val);

                if let Some(left_ref) = node.left.as_ref() {
                    q.push_back(left_ref.clone());
                }

                if let Some(right_ref) = node.right.as_ref() {
                    q.push_back(right_ref.clone());
                }
            }

            if flip_flop {
                temp.reverse();
            }

            res.push(temp);
            flip_flop = !flip_flop;
        }

        res
    }
}

File: algorithms/2_medium/E/01609_even_odd_tree/bfs.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isEvenOddTree(TreeNode* root) {
        int curr_level{ 0 };
        queue<TreeNode*> q{};
        q.push(root);

        while (!q.empty()) {
            const size_t curr_len{ q.size() };
            int prev_val{
                (curr_level & 1) == 1 ? numeric_limits<int>::max() : numeric_limits<int>::min()
            };

            for (size_t _i{ 0 }; _i < curr_len; _i += 1) {
                auto node = q.front();
                q.pop();

                if ((curr_level & 1) == (node->val & 1)) {
                    return false;
                }

                if ((curr_level & 1) == 0) {
                    if (prev_val >= node->val) {
                        return false;
                    }
                } else {
                    if (prev_val <= node->val) {
                        return false;
                    }
                }

                prev_val = node->val;

                if (node->left != nullptr) {
                    q.push(node->left);
                }

                if (node->right != nullptr) {
                    q.push(node->right);
                }
            }

            curr_level += 1;
        }

        return true;
    }
};
